rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,266 @@
+deployment of dynamic tokens.
+
+## Details
+
+Much inspiration is taken from Vault's [IAM auth method](https://www.vaultproject.io/docs/auth/aws#iam-auth-method).
+
+### Authentication
+
+In place of a join token, nodes will present two signed AWS API requests to the
+auth server: `sts:GetCallerIdentity` and `organizations:DescribeOrganization`
+
+These are signed HTTP requests which will be serialized and sent to the auth
+server over GRPC.
+
+The auth server will then:
+1. Check that the request URL is a valid AWS API endpoint.
+2. Send each request to the AWS API.
+  - AWS will check the signature and return an error if it is invalid, in which case the node will be rejected.
+3. Extract the Account from the `sts:GetCallerIdentity` response
+4. Extract the Organization from the `organizations:DescribeOrganization` response
+5. Check the Organization and Account against the configured deny rules. Reject the node if any match.
+6. Check the Organization and Account against the configured allow rules. Reject the node if none match.
+7. Possibly extra checks to prevent replay attacks, see the following section.
+
+By presenting signed AWS API requests, the node proves that it has access to
+AWS credentials. We rely on AWS to verify the signatures.
+
+The API responses prove the Organization and Account of the credentials
+used to sign the requests. We check that the request is a real AWS API
+endpoint, and rely on AWS to return the real identity for the signed
+requests.
+
+Nodes never share their AWS credentials, only signed requests.
+
+### Mitigating the Risk of Replay Attacks
+
+Signed API requests will only be held in memory and never logged or written to
+disk. All transport of signed API requests will be over TLS.
+
+We can include arbitrary headers in the signed API request and AWS will verify
+the signature over those headers. We can use this to include a timestamp on the
+request and enforce a TTL. Amazon does this by default with a TTL of 15
+minutes.  We could set a shorter TTL ourselves, or make it configurable.
+
+We could also include a UUID header, and use this to verify that each signed
+request is only used once by having the auth server store the UUID of requests
+that have been used recently and still have a valid TTL. With multiple auth
+servers this would require some backend coordination.
+
+### Teleport Configuration
+
+Support for new configuration options will be added to Teleport's configuration
+file in order to enable and configure this feature.
+
+```yaml
+teleport:
+
+  # This section should be used on nodes which will join the cluster with the
+  # new aws join method, in place of auth_token.
+  aws_token:
+    enabled: true
+
+auth_service:
+
+  # This section should be used on auth servers which will allow nodes to
+  # join the cluster with the new aws join method.
+  aws_join:
+    # Deny rules will be checked first. If any deny rule matches an incoming
+    # node, it will be rejected.
```

Comment:
Does it imply 
>it will be rejected, if no allow rule set

?

Reply:
yes, if an incoming node does not match any allow rule it will be rejected